analyses/malawi/flooding_trigger/08_mwi_event_visualization.md

```diff
@@ -0,0 +1,130 @@
+sys.path.append(path_mod)
+
+from src.indicators.flooding.config import Config
+from src.indicators.flooding.glofas import utils
+
+config = Config()
+mpl.rcParams['figure.dpi'] = 300
+
+PLOT_DIR = config.DATA_DIR / 'processed' / 'mwi' / 'plots' / 'flooding'
+PRIVATE_DIR = config.DATA_PRIVATE_DIR
+EXPLORE_DIR = PRIVATE_DIR / 'exploration' / 'mwi' / 'flooding'
+
+LEADTIMES = [x + 1 for x in range(10)]
+SAVE_PLOT = False
+EVENT = 'RCO' # 'rco', 'floodscan', 'combined'
+COUNTRY_ISO3 = 'mwi'
+
+stations_adm2 = {
+    #'G1724': 'Nsanje',
+    'G2001': 'Chikwawa'
+}
+```
+
+Read in the processed GloFAS data.
+
+```python
+ds_glofas_reanalysis = utils.get_glofas_reanalysis(
+    country_iso3=COUNTRY_ISO3)
+ds_glofas_reforecast = utils.get_glofas_reforecast(
+    country_iso3 = COUNTRY_ISO3, leadtimes=LEADTIMES,
+    interp=True
+)
+ds_glofas_reforecast_summary = utils.get_glofas_forecast_summary(ds_glofas_reforecast)
+df_return_period = utils.get_return_periods(ds_glofas_reanalysis, method='analytical')
+```
+
+Read in the processed event dataset(s). Here, we're using the combined RCO events from both Chikwawa and Nsanje. 
+
+```python
+events = {}
+for station in stations_adm2.values():
+    events[station] = pd.read_csv(EXPLORE_DIR / f'all_rco_event_summary.csv')
+```
+
+Create a plot for each event. For the dates around each flood onset, we want to see what both the GloFAS reanalysis (historical) and reforecast data look like. We'll pick a specific forecast probability and include forecasts from various leadtimes. The y-axis is scaled to the percentage of a given return period threshold. Lines on the graph have a stronger weight when the values exceed the threshold.
+
+```python
+buffer_start = 5
+buffer_end = 30
+forecast_probability = 70
+rp = 2
+colors = plt.cm.Blues(np.linspace(0.4,1,len(LEADTIMES)))
+
+for code, station, in stations_adm2.items():
+    
+    nevents = len(events[station].index)
+    nrows = 6
+    ncols = round(nevents/nrows)
+    igroup = 1 
+    
+    fig, axs = plt.subplots(nrows, ncols, figsize=(10,10), squeeze=False, sharey=True)
+    
+    for i in range(nrows):
+
+        for j in range(ncols):
+            
+            try: 
+                ax = axs[i,j]
+                ax.axes.xaxis.set_ticks([])
+                ax.set_ylim([0.25, 1.25])
+
+                flood_onset_date = pd.to_datetime(events[station].loc[igroup-1, 'start_date'])
+                glofas_start_date = flood_onset_date - np.timedelta64(buffer_start, 'D')
+                glofas_end_date = flood_onset_date + np.timedelta64(buffer_end, 'D')
+                rp_val = df_return_period.loc[rp, code]
+
+                for ix, leadtime in enumerate(LEADTIMES):
+                    glofas_reforecast_sel = (
+                        ds_glofas_reforecast_summary[code]
+                        .shift(time=-leadtime) # TODO: I'm actually a bit unsure of whether this shift needs to happen or not...
```

Comment:
This can be removed! `utils.get_glofas_reforecast` shifts the dates by default so each date corresponds to whta is being forecasted for that date. (There is also the option to turn it off if needed)

Reply:
![Capture](https://user-images.githubusercontent.com/25990216/129420386-4d304d20-032d-4d91-9ab4-21c7f5ed4ec6.PNG)

So that's what I thought but was a bit uncertain with many of the figures looking like this without the shift in the forecast data... does this also look odd to you? (and like there's still a shift that needs to happen)

Reply:
You're right that this looks very suspicious (actually most of your examples do), I became a bit worried and did a few checks but now I'm sure that the code is working as intended, which is:
- `shift_dates = False` means that the date corresponds to the day the forecast was made
- `shift_dates = True` means that this level of discharge is forecasted for that date

The checks I did included doing some test downloads from the CDS interface and looking at the variable descriptions in the raw data. But I think one clear indicator that we have it correct is the following: the reforecast is run twice per week, and when you get the raw data, it's only populated for two days per week (rather than all dates), which indicates that the date corresponds to the one where the forecast was run. I hope that makes sense.

Another check that I found useful was to use the following code snippet in your notebook:
```python
# Slice time and get mean of ensemble members for simple plotting
start = '2001-01-15'
end = '2001-5-31'
station = 'G1724'

rf_list_slice = ds_glofas_reforecast.sel(time=slice(start,end))[station]
ra_slice = ds_glofas_reanalysis.sel(time=slice(start, end))[station]

rf_list_slice.median(axis=1).plot.line( x='time', add_legend=True, lw=1)
ra_slice.plot.line(label='Historical', c='k', lw=1)
plt.show()
```
If you try with `shift_dates=True` and `False` respectively, you get the following:
![image](https://user-images.githubusercontent.com/6585693/129540428-4e12cd6c-d816-4ea0-85d5-cee127cdfb15.png)
![image](https://user-images.githubusercontent.com/6585693/129540539-913b0170-47e4-4060-b57a-866602ba05bd.png)

So while not shifting the peaks does give a better peak alignment right around 02-01, the rest looks a bit off, particularly once the discharge is decreasing, where I think the forecast should be performing better.

Reply:
Thanks for checking up on that! I'll remove the shift from this notebook. So do you think that the suspicious-looking lags in these plots from ```shift=true``` is just an artefact of the forecast?

Reply:
Yes, I was thinking it could be due to the sub sampling and subsequent interpolation, it may cause the peak to appear as if it is shifting every day. It could be something to ask Ervin though.